File: day23.go
package main

import "fmt"

const n = 1000000
const times = 10000000

func main() {
	data := []int{2, 5, 3, 1, 4, 9, 8, 6, 7}
	datalen := len(data)
	nodes := make([]int, n+1)
	for i := 0; i < n; i++ {
		a := i + 1
		if i < datalen {
			a = data[i]
		}
		b := i + 2
		if (i+1)%n < datalen {
			b = data[(i+1)%n]
		}
		nodes[a] = b
	}
	current := data[0]
	for i := 0; i < times; i++ {
		a := nodes[current]
		b := nodes[a]
		c := nodes[b]
		nodes[current] = nodes[c]
		destination := current
		for {
			destination--
			if destination < 1 {
				destination = n
			}
			if destination != a && destination != b && destination != c {
				break
			}
		}

		nodes[c] = nodes[destination]
		nodes[destination] = a
		current = nodes[current]
	}

	answer := nodes[1] * nodes[nodes[1]]
	fmt.Println(answer)
}

File: day23.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char **argv) {
    int data[] = {2, 5, 3, 1, 4, 9, 8, 6, 7};
    const int datalen = sizeof(data)/sizeof(int);
    const int n = 1000000;
    const int times = 10000000;
    int *nodes = (int *)malloc((n+1) * sizeof(int));
    for (int i = 0; i < n; ++i) {
        int a = i < datalen ? data[i] : i + 1;
        int b = (i + 1) % n < datalen ? data[(i+1) % n] : i + 2;
        nodes[a] = b;
    }

    int current = data[0];

    for (int i = 0; i < times; ++i) {
        int a = nodes[current];
        int b = nodes[a];
        int c = nodes[b];
        nodes[current] = nodes[c];
        int destination = current;
        while (1) {
            destination -= 1;
            if (destination < 1)
                destination = n;
            if (destination != a && destination != b && destination != c)
                break;
        }
        
        nodes[c] = nodes[destination];
        nodes[destination] = a;
        current = nodes[current];
    }

    long answer = (long)nodes[1] * (long)nodes[nodes[1]];
    printf("%lu\n", answer);

    free(nodes);
    return 0;
}
